### ISSUE TITLE
Smooth Ruby upgrades without breakage and residue files

### ISSUE THREAD

**Issue by forthrin** [rubygems/rubygems#7482]
How should Ruby developers handle a major version upgrade of Ruby?

Preamble:

- In the related environment, only a single, latest version is needed.
- When a new version arrives, old versions and dependencies can be deleted immediately.
- Some things may relate to Homebrew, but starting here for expertise since it's all Ruby/Gems.
- Note the following directories:

```
    /opt/homebrew/lib/ruby/gems/3.2.0/bin
    /opt/homebrew/lib/ruby/gems/3.3.0/bin
    /opt/homebrew/opt/ruby/bin
```

Observations & Questions:

1. When Ruby was upgraded to 3.3, all gems had to be reinstalled, because there are separate directories for each version. Is there a one-liner for reinstalling all gems for the new version? Or a way to simply carry over the gems?

2. `lib/ruby/gems/*/bin` and `opt/ruby/bin` are semi-duplicates. What is the logic behind this, and what goes where? Why can't this be a single directory? (Preferably the short non-versioned, that doesn't break with upgrades.)

```
     3.3       rubocop
     3.3       ruby-parse
     3.3       ruby-rewrite
bin            gem
bin            racc
bin            rake
bin            rdbg
bin            rdoc
bin            ri
bin            ruby
bin            syntax_suggest
bin  3.3       bundle
bin  3.3       bundler
bin  3.3       nokogiri
(etc. for all installed gems)
```

3. `ri` keeps an humongous backlog of old versions. A single gem may have nearly 10.000 files worth, to the extent that it slows down `find /opt` noticeably. Can it be configured to exclude versions older than X?

```
$ cd /opt/homebrew/lib/ruby/gems/3.2.0/doc/parser-3.3.0.1/ri
$ find . -type f | wc -l
    9269
$ find . -type f | cut -d / -f 1-3 | uniq -c | grep Ruby
 430 ./Parser/CurrentRuby
 405 ./Parser/MacRuby
 359 ./Parser/Ruby18
 399 ./Parser/Ruby19
 430 ./Parser/Ruby20
 437 ./Parser/Ruby21
 439 ./Parser/Ruby22
 440 ./Parser/Ruby23
 443 ./Parser/Ruby24
 439 ./Parser/Ruby25
 442 ./Parser/Ruby26
 530 ./Parser/Ruby27
 527 ./Parser/Ruby28
 535 ./Parser/Ruby30
 550 ./Parser/Ruby31
 549 ./Parser/Ruby32
 544 ./Parser/Ruby33
 401 ./Parser/RubyMotion
```

**Comment by hsbt** [comment]
>ri keeps an humongous backlog of old versions. A single gem may have nearly 10.000 files worth, to the extent that it slows down find /opt noticeably. Can it be configured to exclude versions older than X?

No, It's not standard case. They are only for `parser` gem.

**Comment by forthrin** [comment]
Admittedly, it's the top one. Still even for my modest collection of Gems, there are 50k+ doc files. 100k+ if you count Ruby 3.2, which I intend to get rid of by some more civilized way than `rm -rf` if someone helps with the questions.

```
$cd /opt/homebrew/lib/ruby/gems/3.2.0/doc
$ find . -type f | wc -l
   54951
$ find . -type f | cut -d / -f 1-3 | uniq -c | sort -nr | head
9269 ./parser-3.3.0.1/ri
7291 ./rubocop-1.59.0/ri
2500 ./language_server-protocol-3.17.0.3/ri
2001 ./rubygems-3.5.4/ri
1890 ./rbs-3.4.1/ri
1500 ./httpx-1.2.0/ri
1382 ./concurrent-ruby-1.2.2/ri
1371 ./rdoc-6.6.2/ri
1337 ./pry-0.14.2/ri
1286 ./rss-0.3.0/ri
```

**Comment by hsbt** [comment]
You can ignore to generate docs from `gem install` with `--no-document` option.

```
$ cat ~/.config/gem/gemrc
gem: "--no-document"
```

**Comment by deivid-rodriguez** [comment]
I think 3rd remark has been addressed.

Regarding the first, there's no built-in way to carry over gems across Ruby versions because gems with extensions need to be recompiled against the new Ruby. There's no builtin way either to reinstall all gems from one Ruby to another but you could automate that with a script that parses `gem list` run using the old ruby, and installs the result into the new Ruby.

Regarding the second question, I'm not fully sure why those are duplicated. I think you could ask in the homebrew project about their specific layout.

**Comment by forthrin** [comment]
Very nice answer. I didn't even notice that Ruby was bumped a major version because that would drown in screens of other terminal output when running `brew upgrade`. When a major version upgrades happens, several things must be done:

- Read the change log on the Ruby webpage and update own code if necessary
- Update `$PATH` (and possibly others)
- Reinstall gems (good point that binaries must be recompiled)
- Delete everything belonging to the old version

It would be nice if this was automated in some way, since everyone have to go through the same steps. Not sure where it would belong. Suggestions?

**Comment by deivid-rodriguez** [comment]
Those sounds like some automations that `brew upgrade` should be taking care of?

**Comment by forthrin** [comment]
I'll check next time. It possibly does some of the things, but like I said, with the screenfuls of verbose output that comes with a regular `brew upgrade`, you're bound to miss it. `brew upgrade` should have had some sort of `--highlight-major-changes`option.

**Comment by deivid-rodriguez** [comment]
My understanding is that these automations would make it so that things just work after new Ruby is installed, so it would not be a big deal if you miss the output?

In any case, RubyGems is not a Ruby installer, so this kind of thing feels out of scope of RubyGems to me.

**Comment by forthrin** [comment]
Good point. I'll consider this again for the next major upgrade that leaves residue file and unclear/broken configurations.

Out of curiosity: How, specifically, do do you manage versioning personally? (Maybe not the right person to ask if you're so close to the core that you have full overview and control :)

**Comment by deivid-rodriguez** [comment]
From my experience, most rubyists manage their Rubies using a ruby version manager like RVM, Rbenv, or chruby. I used those for a long time and they work fine. I did recently switch to asdf for convenience (I had to work with several different languages and it was handy to use the same manager for all).

I'm closing this because I don't think there's actionable stuff for us in RubyGems, but let me know if you disagree!
